src/Features/Core/Portable/CodeFixes/Configuration/ConfigureSeverityLevelCodeFixProvider.ConfigurationUpdater.cs

```diff
@@ -0,0 +1,357 @@
+            /// so that the severity of the given <paramref name="diagnostic"/> is configured to be the given
+            /// <paramref name="severity"/>.
+            /// </summary>
+            public static async Task<Solution> ConfigureEditorConfig(
+                string severity,
+                Diagnostic diagnostic,
+                Project project,
+                CancellationToken cancellationToken)
+            {
+                var updater = new ConfigurationUpdater(severity, diagnostic, project, cancellationToken);
+                return await updater.Configure().ConfigureAwait(false);
+            }
+
+            private async Task<Solution> Configure()
+            {
+                var solution = _project.Solution;
+
+                // Find existing .editorconfig or generate a new one if none exists.
+                var editorConfigDocument = FindOrGenerateEditorConfig(ref solution);
+                if (editorConfigDocument == null)
+                {
+                    return solution;
+                }
+
+                var result = await editorConfigDocument.GetTextAsync(_cancellationToken).ConfigureAwait(false);
+                var headers = new Dictionary<string, TextLine>();
+
+                // For option based code style diagnostic, try to find the .editorconfig key-value pair for the
+                // option setting.
+                var (optionNameOpt, optionValueOpt) = FindEditorConfigOptionKeyValueForRule();
+
+                // Check if an entry to configure the rule severity already exists in the .editorconfig file.
+                // If it does, we update the existing entry with the new severity.
+                var configureExistingRuleText = CheckIfRuleExistsAndReplaceInFile(optionNameOpt, result, headers);
+                if (configureExistingRuleText != null)
+                {
+                    return solution.WithAnalyzerConfigDocumentText(editorConfigDocument.Id, configureExistingRuleText);
+                }
+
+                // We did not find any existing entry in the in the .editorconfig file to configure rule severity.
+                // So we add a new configuration entry to the .editorconfig file.
+                var addMissingRuleText = AddMissingRule(optionNameOpt, optionValueOpt, result, headers);
+                if (addMissingRuleText != null)
+                {
+                    return solution.WithAnalyzerConfigDocumentText(editorConfigDocument.Id, addMissingRuleText);
+                }
+
+                // Bail out and return the old solution for error cases.
+                return solution;
+            }
+
+            private AnalyzerConfigDocument FindOrGenerateEditorConfig(ref Solution solution)
+            {
+                if (_project.AnalyzerConfigDocuments.Any())
+                {
+                    var diagnosticFilePath = PathUtilities.GetDirectoryName(_diagnostic.Location.SourceTree?.FilePath ?? _project.FilePath);
```

Comment:
these might both be null.  Is GetDirectoryName and IsAbsolute resilient to that?

Reply:
Yes.